src/storage/src/sink/kafka.rs

```diff
@@ -696,6 +696,19 @@ impl KafkaSinkState {
                         latest_ts = Some(ts);
                     }
                 }
+
+                // If the next possible offset for the client is past the high watermark, we've seen
+                // everything we expect to see.
+                let position = progress_client
+                    .position()?
+                    .find_partition(progress_topic, partition)
+                    .ok_or_else(|| anyhow!("No progress info for known partition"))?
+                    .offset();
+                if let Offset::Offset(i) = position {
+                    if i >= hi {
```

Comment:
what happens if its actually greater, not equal to? To me it seems like that would be incorrect unless we fence the producer (where do we do that, btw?)

Reply:
> what happens if its actually greater, not equal to?

Maybe the surprise here is that there can be "gaps" in the offsets because of failed transactions. So you might get, for example:
```
0. u1: t0
1. u2: t0
2. <missing>
3. <missing>
4. u1: t1
```

If I'm the sink for `u2`, I might start up with `lo=0` and `hi=3`, and then as i'm scanning through the offsets i'll skip straight from a message with offset 1 to a message with offset 4. I don't think there's any point where `position` will return `3`.

> where do we do that, btw?

The call to `init_transactions` should do the fencing IIRC, which happens just before `determine_latest_progress_record`.